### PR TITLE
vochain: refactor vote transaction check

### PR DESCRIPTION
**Pull request by p4u**
see commit messages

### CODE REVIEW

File: vochain/transaction/vote_tx.go [hunk 1]
```diff
@@ -14,25 +14,35 @@ import (
 
 // VoteTxCheck performs basic checks on a vote transaction.
 func (t *TransactionHandler) VoteTxCheck(vtx *vochaintx.VochainTx, forCommit bool) (*vstate.Vote, error) {
```
Comment (p4u) [inline vochain/transaction/vote_tx.go:16]: This is the method where I'd need review. It is a very crucial part of the system and the logic should be double-checked. @mvdan @altergui @lucasmenendez @jordipainan

File: vochain/transaction/vote_tx.go [hunk 2]
```diff
@@ -43,69 +53,86 @@ func (t *TransactionHandler) VoteTxCheck(vtx *vochaintx.VochainTx, forCommit boo
...
-
-		// if vote is in cache, lazy check
-		if vote != nil {
-			// if not forCommit, it is a mempool check,
-			// reject it since we already processed the transaction before.
-			if !forCommit {
-				return nil, ErrorAlreadyExistInCache
-			}
+	// Check if the vote is already in the cache
+	// In order to avoid double vote check (on checkTx and delivoteEnveloperTx), we use a memory vote cache.
```
Comment (jordipainan) [inline vochain/transaction/vote_tx.go]: Typo `delivoteEnveloperTx` to `deliverTx`

File: vochain/transaction/vote_tx.go [hunk 3]
```diff
@@ -43,69 +53,86 @@ func (t *TransactionHandler) VoteTxCheck(vtx *vochaintx.VochainTx, forCommit boo
...
+	// An element can only be added to the vote cache during checkTx.
+	// EvoteEnvelopery N seconds the old votes which are not yet in the blockchain will be removoteEnveloped from cache.
```
Comment (jordipainan) [inline vochain/transaction/vote_tx.go]: Typo `EvoteEnvelopery` to `Every`

**Review by jordipainan — APPROVED** [review]

**Comment by altergui** [comment]
LGTM, i didn't do my review in time, sorry about that, but here's a small (late) contribution: what testsuite covers of VoteTxCheck (i.e. tested on each CI run). the only significant omission is	"// if vote exists, check if it has reached the max overwrite count" this is not covered. but it's precisely what Maria is working on.


image
image
image

**Pull request merged**